### ISSUE TITLE
stat: "running" subpackage for generator-like statistics?

### ISSUE THREAD

**Issue by btracey** [gonum/gonum#1037]
Just a thought -- logging as an issue so I don't forget / can solicit feedback.

At the moment we have `stat.Mean(x, weights []float64) float64` which computes the mean of a set of data. The thought is to also provide something like stat/running

```
package running

type Mean struct {
    count float64
    mean float64
}

func (m *Mean) Mean() float64 {
    return m.mean
}

func (m *Mean) Count() float64 {
    return m.count 
}

func (m *Mean) Add(x float64) {
      m.mean = x + m.mean * m.count / (m.count + 1)
      m.count++
}

func (m *Mean) AddWeighted(x, weight float64){ .... }

func (m *Mean) Reset() { ... }
```

and there could be similar for Variance/StdDev, DiscountedMean (where the count is discounted by a constant, etc.). These could be relatively easily combined with a shim:

```
type MeanStd struct {
    running.Mean()
    running.StdDev()
}

func (m *MeanStd) Add(x float64) {
    m.Mean.Add(x)
    m.StdDev.Add(x)
}
```

This would help both with streams that are too large to want to put in a `[]float64` and data that comes from some source (say, a channel) and so cannot be put into a `float64`. 

I don't think `Add` is the right verb, but I'm not sure what the right verb is (`Accumulate` seems too heavy, `Append` also seems wrong ...)

**Comment by kortschak** [comment]
I have often thought that this would be useful. Also note that you can calculate the variance in the same loop very cheaply.

```
// The Method of Provisional Means
n++
mean = oldmean + (x-oldmean)/n
sumOfSquareDiffs += (x - oldmean) * (x - mean)
oldmean = mean
```

Then use the `sumOfSquareDiff` as you would expect.

**Comment by btracey** [comment]
Yes, agreed, and since it's so common we'd probably want a specific MeanVariance type (or maybe just one thing that's the kitchen sink, i.e. `running.Stats`). I was just trying to mention that it's easy to combine different things together so the combinatorics aren't so bad.
